### DISCUSSION TITLE
Possible StreamingResponse performance issues with large videos

### DISCUSSION THREAD

**Discussion by LindezaGrey** [fastapi/fastapi#6949] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Question

Hi everyone!
I would like to know if i would have to expect performance issues from FastAPI itself, when i implement the following (given that my server is big enough):
FastAPI is responding to an API call with a StreamingResponse and streams a large video file (e.g. 1GB .webm) to roughly 200 clients simultaneously. Is there anything that i have to take care of programmatically or from architectural side? If you have doubt that it will work i would switch to an additional dedicated video streaming backend. 
Has anybody tried something similiar?
Thanks in advance for your answers :)

**Comment by falkben** [comment]
Streaming video is notoriously complicated. Have you thought of offering different bitrates? How about different compression types (not all browsers are able to view the same)? What happens if performance degrades -- how do your clients get resynched? Video files are also typically served as smaller chunks (so not serving just one large file but smaller chunks at different compression rates). Are you going to be breaking the files up yourself?

My understanding is that most people rely on client side video players for this, and then serve content from something like AWS.  Or just use a service...  You could just offer up static files using nginx or something else, but your concern there is going to be with bandwidth. Static content serving is not CPU intensive so I don't think you have to worry about that. You could implement the API call in FastAPI and just direct the user to a stream hosted on a cloud provider for instance.

Hope this helps.

**Comment by LindezaGrey** [comment, marked as answer]
Thank you for your answer, and sorry for not responding so quickly. I evaluated several options, and here are my findings, if anyone is facing a similar usecase:
- Streaming directly through FastAPI doesn't make much sense and was hardly working
- I used a RedirectResponse in FastAPI to redirect to a nginx server who was just serving the static videofiles, performance wasn't a problem up to roughly 50 concurrent viewers on the same video file (20minutes long, 200mb) when the Bandwidth of my VM became a bit overwhelmed (1gbit upstream connection). My frontend was using plyr as media player.
-  now using a RedirectResponse to redirect to a video stream CDN, which also handles encoding to various quality settings automatically

**Comment by falkben** [comment]
Thanks for coming back to close the issue and reporting what you ended up doing!

Just note, if the files are being served by your nginx server (not the CDN, as you currently have setup), you don't have to do a redirect response, you can set the `x-accel-redirect` header and return that, and, if you have x-accel enabled w/ your nginx server, it will use that to serve the file. This is preferable because your client doesn't see the redirect, the "redirect" is done internally on your backend, and should be faster. More info here: https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/

**Comment by tiangolo** [comment]
Thanks for the help here @falkben ! 👏  🙇 

Thanks for reporting back and closing the issue @LindezaGrey  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
